### DISCUSSION TITLE
Modeling structures that are mostly (but not completely) records

### DISCUSSION THREAD

**Discussion by aabs** [microsoft/typespec#3310] [Q&A] [answered]
Hi, Apologies for the noobie question. I'm still trying to find my feet with typespec. 

I'm dealing with a 3rd party telemetry API that delivers responses like this, and I'm unsure how to model it with typespec:

![image](https://github.com/microsoft/typespec/assets/157775/9e59d346-d163-4b79-bdaa-097d56c4a949)

I figure I can use a `Record<>` to represent the contents of `result_data`, but in every case the items that sit under record fields like `1370462_1_1_1` are a whole bunch of different fields with keys (starting with `p`), plus a `time_stamp` field.  I guess I could just go ahead and make that a `Record<string>` and then cast to `double` or `DateTime` on receipt.  

Is this possible/acceptable?

```
model AssetTelemetry {
  ...Record<string>;
  time_stamp: string;
}
```

How could I model a structure like this (idiomatically)?  FWIW, In practice this will be a very large payload with many records, and many data points beneath the records.

**Comment by bterlson** [comment, marked as answer]
Yeah spreading of a record is acceptable and is the preferred way to represent this kind of thing where you have a known properties and then a bunch of unknown properties. So to me that looks like:

```
model Result {
   result_code: string;
   // etc.
   result_data: Record<ResultData>
}

model ResultData {
  ... Record<string>;
  
  time_stamp: string;
}
```

That is a strange timestamp format I have never seen before. Does it have a name?

**Reply by aabs** [reply]
Thanks for the confirmation.  This data is accessed via the `AdditionalData` Property by the look of it?  Is there a way to ensure the known field `time_stamp` is accessible via a property of its own? It doesn't show up when I use the spread code above.

**Reply by bterlson** [reply]
Where are you observing the property isn't showing up? It seems to be doing the right thing for e.g. in [openapi](https://typespec.io/playground?c=bW9kZWwgUmVzdWx0IHsNCiAgIHLFDV9jb2RlOiBzdHJpbmc7xRkvLyBldGMuzCVkYXRhOiBSZWNvcmQ8xkZEYXRhPg0KfQ0KDQrMXsQYxmIuLi7IM8ZfPsVgxAR0aW1lX3N0YW1wy3t9&e=%40typespec%2Fopenapi3&options=%7B%7D).

**Reply by aabs** [reply]
Sorry, I didn't mention that I'm trying to reverse engineer the OpenAPI spec of a 3rd party that doesn't get the 'open' part of OpenAPI, and are not sharing their spec.  So in practice I am just feeding the resulting OpenAPI file directly into Kiota to create a client to access the API.  

Here's the Makefile I use:

```
.PHONY: all client

all: tsp-output\@typespec\openapi3\openapi.yaml

tsp-output\@typespec\openapi3\openapi.yaml: main.tsp
	tsp compile . --pretty

client: tsp-output\@typespec\openapi3\openapi.yaml
	 kiota generate --exclude-backward-compatible -l CSharp -d "tsp-output\@typespec\openapi3\openapi.yaml" -o Client -c SungrowOpenApiClient -n Sungrow  --co
```

**Reply by aabs** [reply]
So, I'm not seeing it in the Kiota generated code.

![image](https://github.com/microsoft/typespec/assets/157775/48004af9-4726-43e6-b1db-8039fb1d83d0)

**Reply by bterlson** [reply]
Ah, I would guess this is a Kiota limitation then. Maybe @baywet could confirm? You could also file an issue there.

**Reply by baywet** [reply]
Thanks for looping me in. @aabs Can you share the resulting OpenAPI section please?
@bterlson are you mapping Record<string> to an OAS patterned property in the conversion? what was the intent here?

**Reply by bterlson** [reply]
@baywet we aren't mapping to pattern property but using `additionalProperties`. You can see an example output [here](https://typespec.io/playground?c=bW9kZWwgUmVzdWx0IHsNCiAgIHLFDV9jb2RlOiBzdHJpbmc7xRkvLyBldGMuzCVkYXRhOiBSZWNvcmQ8xkZEYXRhPg0KfQ0KDQrMXsQYxmIuLi7IM8ZfPsVgxAR0aW1lX3N0YW1wy3t9&e=%40typespec%2Fopenapi3&options=%7B%7D).

**Reply by aabs** [reply]
Hi @baywet, here's the relevant sections:

```yaml
  schemas:
    AssetTelemetryRecord:
      type: object
      required:
        - time_stamp
      properties:
        time_stamp:
          type: string
      additionalProperties:
        type: string
    MetricsResponse:
      type: object
      required:
        - req_serial_num
        - result_code
        - result_msg
        - result_data
      properties:
        req_serial_num:
          type: string
        result_code:
          type: string
        result_msg:
          type: string
        result_data:
          type: object
          additionalProperties:
            type: array
            items:
              $ref: '#/components/schemas/AssetTelemetryRecord'
```

Is this enough detail?

**Reply by aabs** [reply]
> That is a strange timestamp format I have never seen before. Does it have a name?

@bterlson Don't get me started on this topic!  I suspect they have never heard of ISO8601.

**Reply by bterlson** [reply]
It's such a near miss, just add a `T` and a `Z` (assuming this represents UTC) and it's ISO8601 basic format 😁

**Reply by aabs** [reply]
UTC? What's that? 😆

**Reply by baywet** [reply]
Thanks for the additional information. Kiota adds an additional data dictionary/record (string:object) property to models, but it's a convention from kiota itself. It doesn't do anything with the additional properties information from the OAS description.
I hope that clarifies things!

**Reply by bterlson** [reply]
@baywet that means, I think, that while the additionaProperties openapi is effectively ignored, the API should be usable regardless via the additional data dictionary?

**Reply by baywet** [reply]
correct

**Comment by aabs** [comment]
> Thanks for the additional information. Kiota adds an additional data dictionary/record (string:object) property to models, but it's a convention from kiota itself. It doesn't do anything with the additional properties information from the OAS description. I hope that clarifies things!

Do you mean that, because there is a catch-all definition in there, the explicitly named property is ignored?

**Reply by baywet** [reply]
To be more specific, this is what I'm saying is ignored.

```yaml
additionalProperties:
        type: string

additionalProperties:
            type: array
            items:
              $ref: '#/components/schemas/AssetTelemetryRecord'
```

The other properties/types should be generated for you.

**Reply by aabs** [reply]
Thanks.  That is a little problematic for me when it comes to deserializing elements of type `AssetTelemetryRecord`.  This definition is the only use of that type, so it doesn't get generated by default. 
Of course I can create some sort of Dummy API endpoint that references the type, so I have it around, but it feels contrived to do that.

**Reply by baywet** [reply]
We could imagine adding support for projecting additional properties schema to a type, and make sure it does not get trimmed by the generator because it's "unused". This would be a stopgap to supporting pattern properties, once the parsing library we use actual supports OAS 3.1.
Can you create an issue at https://github.com/microsoft/kiota please?
